src/sync/scheduler.ts

```diff
@@ -0,0 +1,146 @@
+import { Task, TaskType } from "~/src/sync/sync.types";
+import { RepoSyncState } from "models/reposyncstate";
+import { getTargetTasks } from "~/src/sync/installation";
+import { createInstallationClient } from "utils/get-github-client-config";
+import Logger from "bunyan";
+import { booleanFlag, BooleanFlags } from "config/feature-flags";
+
+const getCursorKey = (type: TaskType) => `${type}Cursor`;
+const getStatusKey = (type: TaskType) => `${type}Status`;
+
+// These numbers were obtained experimentally by syncing a large customer with 60K quota from GitHub
+const RATE_LIMIT_QUOTA_PER_TASK_RESERVE = 500;
+const MAX_NUMBER_OF_SUBTASKS = 100;
+const SUBTASKS_POLL_MAX_SIZE = MAX_NUMBER_OF_SUBTASKS * 10;
+
+/**
+ *
+ * @param subscription
+ * @param logger
+ * @param mainTask
+ * @param otherTasks - the array will be shuffled after the call, BEWARE!
+ */
+const calculateTasksUsingGitHubRateLimitQuota = async (subscription: Subscription, logger: Logger, mainTask: Task, otherTasks: Task[]) => {
+	try {
+		const metrics = {
+			trigger: "ratelimit_check_backfill"
+		};
+		const rateLimitResponse = await (await createInstallationClient(subscription.gitHubInstallationId, subscription.jiraHost, metrics, logger, subscription.gitHubAppId)).getRateLimit();
+		const rateLimitData = rateLimitResponse.data;
+
+		const availQuota = Math.min(rateLimitData.resources.core.remaining, rateLimitData.resources.graphql.remaining);
+		// We need to reserve some quota for the main task, too
+		const availQuotaForSubtasks = Math.max(0, availQuota - RATE_LIMIT_QUOTA_PER_TASK_RESERVE);
+		const allowedSubtasks = Math.floor(availQuotaForSubtasks / RATE_LIMIT_QUOTA_PER_TASK_RESERVE);
+
+		const nSubTasks = Math.min(allowedSubtasks, MAX_NUMBER_OF_SUBTASKS);
+
+		logger.info({ nSubTasks, rateLimitData }, "Using subtasks: " + nSubTasks);
+
+		if (nSubTasks === 0) {
+			return [mainTask];
+		}
+
+		return [mainTask, ...(
+			otherTasks.slice(0, nSubTasks).sort(() => Math.random() - 0.5)
+		)];
+	} catch (err) {
+		logger.warn({ err }, "Cannot determine rate limit, return only main task");
+		return [mainTask];
+	}
+};
+
+const mapSyncStateToTasks = (tasks: TaskType[], syncState: RepoSyncState): Task[] => {
+	const ret: Task[] = [];
+	tasks.forEach(
+		(taskType) => {
+			if (!syncState[getStatusKey(taskType)] || syncState[getStatusKey(taskType)] === "pending") {
+				ret.push({
+					task: taskType,
+					repositoryId: syncState.repoId,
+					repository: {
+						id: syncState.repoId,
+						name: syncState.repoName,
+						full_name: syncState.repoFullName,
+						owner: { login: syncState.repoOwner },
+						html_url: syncState.repoUrl,
+						updated_at: syncState.repoUpdatedAt?.toISOString()
+					},
+					cursor: syncState[getCursorKey(taskType)] || undefined
+				});
+			}
+		}
+	);
+	return ret;
+};
+
+/**
+ *
+ * @param subscription
+ * @param targetTasks - the first task is always same/deterministic! The tail is random from not complete ones tasks.
```

Comment:
What's the reason for combining JSdocs with TS types? When we migrated from JS -> TS we made an effort to do away with all JSDocs. The additional info for `targetTasks` and following comments are great, but not sure we need to list everything here when we're already typing these in the function signature. Same goes for `calculateTasksUsingGitHubRateLimitQuota`

Reply:
Otherwise IDEA will be complaining....

Reply:
<img width="1515" alt="Screenshot 2023-04-04 at 11 12 29 am" src="https://user-images.githubusercontent.com/37155488/229660670-50504f03-6b0c-4d48-8708-c62420404d00.png">

Just removed the additional params (with no comment) from `calculateTasksUsingGitHubRateLimitQuota` and committed and didn't run into any issues.